dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/ApiTokenController.java

```diff
@@ -0,0 +1,252 @@
+            .addObject( apiToken );
+
+        final ObjectReport objectReport = importService.importMetadata( params ).getFirstObjectReport();
+        final String uid = objectReport.getUid();
+
+        WebMessage webMessage = objectReport( objectReport );
+        if ( webMessage.getStatus() == Status.OK )
+        {
+            webMessage.setHttpStatus( HttpStatus.CREATED );
+            webMessage.setLocation( getSchema().getRelativeApiEndpoint() + "/" + uid );
+
+            // Set our custom web response object that includes the new
+            // generated key.
+            webMessage.setResponse( new ApiTokenCreationResponse( uid, rawKey ) );
+        }
+        else
+        {
+            webMessage.setStatus( Status.ERROR );
+        }
+
+        return webMessage;
+    }
+
+    @Override
+    protected void prePatchEntity( ApiToken newToken )
+    {
+        final ApiToken oldToken = apiTokenService.getWithUid( newToken.getUid() );
+        newToken.setKey( oldToken.getKey() );
+        validateApiKeyAttributes( newToken );
+    }
+
+    @Override
+    protected void prePatchEntity( ApiToken oldToken, ApiToken newToken )
+    {
+        newToken.setKey( oldToken.getKey() );
+        validateApiKeyAttributes( newToken );
+    }
+
+    @Override
+    protected void preUpdateEntity( ApiToken oldToken, ApiToken newToken )
+    {
+        newToken.setKey( oldToken.getKey() );
+        validateApiKeyAttributes( newToken );
+    }
+
+    protected void validateBeforeCreate( ApiToken token )
+    {
+        validateApiKeyAttributes( token );
+    }
+
+    private void validateApiKeyAttributes( ApiToken token )
+    {
+        if ( token.getIpAllowedList() != null )
+        {
+            token.getIpAllowedList().getAllowedIps().forEach( this::validateIp );
+        }
+        if ( token.getMethodAllowedList() != null )
+        {
+            token.getMethodAllowedList().getAllowedMethods().forEach( this::validateHttpMethod );
+        }
+        if ( token.getRefererAllowedList() != null )
+        {
+            token.getRefererAllowedList().getAllowedReferrers().forEach( this::validateReferrer );
+        }
+    }
+
+    private void validateHttpMethod( String httpMethodName )
+    {
+        final ImmutableList<String> validMethods = ImmutableList.of( "GET", "POST", "PATCH", "PUT" );
+        if ( !validMethods.contains( httpMethodName.toUpperCase( Locale.ROOT ) ) )
+        {
+            throw new IllegalArgumentException( "Not a valid http method, value=" + httpMethodName );
+        }
+    }
+
+    private void validateIp( String ip )
+    {
+        InetAddressValidator validator = new InetAddressValidator();
+        if ( !validator.isValid( ip ) )
+        {
+            throw new IllegalArgumentException( "Not a valid ip address, value=" + ip );
+        }
+    }
+
+    private void validateReferrer( String referrer )
```

Comment:
The presence of a referrer header would suggest that the API is being used from a browser, but that's something we want to discourage right? So maybe instead this test should fail if a referrer is present.

We might also want to fail if a user-agent header is present.

Reply:
Ok, I do agree this API tokens might not be well suited for use in a browser context. I think we should discuss it with the team and, either, allow it or do as you say, reject all requests that have the "referer" header.